File: java/roman-numerals/src/main/java/RomanNumerals.java
enum RomanNumerals {
    M(1000), CM(900), D(500), CD(400), C(100), XC(90),
    L(50), XL(40), X(10), IX(9), V(5), IV(4), I(1);

    private final int value;

    RomanNumerals(final int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }

    public static String arabicToRoman(int number) {
        final var sb = new StringBuilder();
        for (var romanNumber : RomanNumerals.values()) {
            while (romanNumber.getValue() <= number) {
                number -= romanNumber.getValue();
                sb.append(romanNumber.toString());
            }
        }
        return sb.toString();
    }
}

class RomanNumeral {
    final private int value;

    public RomanNumeral(int value) {
        this.value = value;
    }

    public String getRomanNumeral() {
        return RomanNumerals.arabicToRoman(value);
    }
}

File: go/roman-numerals/roman_numerals.go
// Package romannumerals provides methods for manipulating numbers into roman numerals
package romannumerals

import "errors"

type RomanNumeral struct {
	Value  int
	Symbol string
}

// RomanNumerals is the mapping between numbers and their corresponding Roman numerals
var RomanNumerals = []RomanNumeral{
	{1000, "M"},
	{900, "CM"},
	{500, "D"},
	{400, "CD"},
	{100, "C"},
	{90, "XC"},
	{50, "L"},
	{40, "XL"},
	{10, "X"},
	{9, "IX"},
	{5, "V"},
	{4, "IV"},
	{1, "I"},
}

// ToRomanNumeral converts an integer to its roman numeral string equivalent
func ToRomanNumeral(number int) (string, error) {
	if number < 1 || number > 3999 {
		return "", errors.New("number has to be in range [1..3999]")
	}
	roman := ""
	for _, numeral := range RomanNumerals {
		for number >= numeral.Value {
			number -= numeral.Value
			roman += numeral.Symbol
		}
	}
	return roman, nil
}
